src/app/tests/TestGroupDataProvider.cpp

```diff
@@ -0,0 +1,408 @@
+
+    err = groups->AddGroupMapping(fabric, group1b, "Group 1.2");
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == err);
+
+    exists = groups->ExistsGroupMapping(fabric, group1a);
+    NL_TEST_ASSERT(apSuite, exists);
+
+    exists = groups->ExistsGroupMapping(fabric, group1b);
+    NL_TEST_ASSERT(apSuite, exists);
+
+    exists = groups->ExistsGroupMapping(fabric, group1c);
+    NL_TEST_ASSERT(apSuite, exists);
+
+    err = groups->RemoveGroupAllMappings(fabric, 1);
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == err);
+
+    exists = groups->ExistsGroupMapping(fabric, group1a);
+    NL_TEST_ASSERT(apSuite, !exists);
+
+    exists = groups->ExistsGroupMapping(fabric, group1b);
+    NL_TEST_ASSERT(apSuite, !exists);
+
+    exists = groups->ExistsGroupMapping(fabric, group1c);
+    NL_TEST_ASSERT(apSuite, !exists);
+}
+
+void TestEndpointIterator(nlTestSuite * apSuite, void * apContext)
+{
+    GroupDataProvider * groups              = GetGroupDataProvider();
+    chip::FabricIndex fabric                = 1;
+    GroupDataProvider::GroupMapping group1a = { .endpoint = 1, .group = 1 };
+    GroupDataProvider::GroupMapping group1b = { .endpoint = 1, .group = 3 };
+    GroupDataProvider::GroupMapping group2a = { .endpoint = 2, .group = 2 };
+    GroupDataProvider::GroupMapping group3a = { .endpoint = 3, .group = 1 };
+    GroupDataProvider::GroupMapping group3b = { .endpoint = 3, .group = 2 };
+    GroupDataProvider::GroupMapping group3c = { .endpoint = 3, .group = 3 };
+    CHIP_ERROR err                          = CHIP_NO_ERROR;
+
+    NL_TEST_ASSERT(apSuite, groups);
+
+    err = groups->AddGroupMapping(fabric, group3b, "Group 3.2");
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == err);
+
+    err = groups->AddGroupMapping(fabric, group2a, "Group 2.2");
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == err);
+
+    err = groups->AddGroupMapping(fabric, group1b, "Group 1.3");
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == err);
+
+    err = groups->AddGroupMapping(fabric, group3a, "Group 3.1");
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == err);
+
+    err = groups->AddGroupMapping(fabric, group1a, "Group 1.1");
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == err);
+
+    err = groups->AddGroupMapping(fabric, group3c, "Group 3.3");
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == err);
+
+    err = groups->AddGroupMapping(fabric, group3a, "Group 3.1");
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == err);
+
+    EndpointId endpoint                          = 1;
+    GroupDataProvider::GroupMappingIterator * it = groups->IterateGroupMappings(fabric, endpoint);
+    NL_TEST_ASSERT(apSuite, it);
+
+    // Endpoint 1
+    uint16_t count1 = it->Count();
+    uint16_t count2 = 0;
+    NL_TEST_ASSERT(apSuite, 2 == count1);
+    while (it->HasNext())
+    {
+        count2++;
+        GroupId gid = it->Next();
+        NL_TEST_ASSERT(apSuite, 1 == gid || 3 == gid);
+    }
+    NL_TEST_ASSERT(apSuite, count2 == count1);
+    delete it;
```

Comment:
Cannot merge until delete no longer needed. Iterator must be refactored to internally manage its memory, such that it can be copied and/or passed by value.